File: 3. Behavioral Design Patterns/Template Method/Typescript/TemplateMethodTemplate.ts
// Template Method Pattern – TypeScript Template

abstract class AbstractClass {
  templateMethod(): void { this.stepOne(); this.stepTwo(); this.hook() }
  protected abstract stepOne(): void
  protected abstract stepTwo(): void
  protected hook(): void { /* optional override */ }
}

class ConcreteClassA extends AbstractClass {
  protected stepOne(): void { console.log('A: Step 1') }
  protected stepTwo(): void { console.log('A: Step 2') }
}

class ConcreteClassB extends AbstractClass {
  protected stepOne(): void { console.log('B: Step 1') }
  protected stepTwo(): void { console.log('B: Step 2') }
  protected hook(): void { console.log('B: Hook') }
}

;(function main() { new ConcreteClassA().templateMethod(); new ConcreteClassB().templateMethod() })()

export {}

File: 3. Behavioral Design Patterns/Template Method/Kotlin/TemplateMethodTemplate.kt
// Template Method Pattern – Kotlin Template

abstract class AbstractClass {
    fun templateMethod() { stepOne(); stepTwo(); hook() }
    protected abstract fun stepOne()
    protected abstract fun stepTwo()
    protected open fun hook() { /* optional override */ }
}

class ConcreteClassA : AbstractClass() {
    override fun stepOne() = println("A: Step 1")
    override fun stepTwo() = println("A: Step 2")
}

class ConcreteClassB : AbstractClass() {
    override fun stepOne() = println("B: Step 1")
    override fun stepTwo() = println("B: Step 2")
    override fun hook() = println("B: Hook")
}

fun main() { ConcreteClassA().templateMethod(); ConcreteClassB().templateMethod() }
